src/rust/engine/ui/src/console_ui.rs

```diff
@@ -198,40 +109,299 @@ impl ConsoleUI {
-struct Instance {
-  tasks_to_display: IndexMap<SpanId, (String, Option<Duration>)>,
-  multi_progress_task: MultiProgressTask,
-  bars: Vec<ProgressBar>,
+pub enum Instance {
+  Indicatif(IndicatifInstance),
+  Prodash(ProdashInstance),
+}
+
+impl Instance {
+  ///
+  /// Setup an Instance of the UI renderer.
+  ///
+  /// NB: This method must be very careful to avoid logging. Between the point where we have taken
+  /// exclusive access to the Console and when the UI has actually been initialized, attempts to
+  /// log from this method would deadlock (by causing the method to wait for _itself_ to finish).
+  ///
+  pub fn new(renderer: &String, local_parallelism: usize, executor: Executor) -> Result<Instance, String> {
+    return match renderer.as_str() {
+      "indicatif-spinner" => {
+        // We take exclusive access to stdio by registering a callback that is used to render stderr
+        // while we're holding access. See the method doc.
+        let stderr_dest_bar: Arc<Mutex<Option<WeakProgressBar>>> = Arc::new(Mutex::new(None));
+        // We acquire the lock before taking exclusive access, and don't release it until after
+        // initialization. That way, the exclusive callback can always assume that the destination
+        // is initialized (i.e., can `unwrap` it).
+        let mut stderr_dest_bar_guard = stderr_dest_bar.lock();
+        let (term_read, _, term_stderr_write) = {
+          let stderr_dest_bar = stderr_dest_bar.clone();
+          stdio::get_destination().exclusive_start(Box::new(move |msg: &str| {
+            // Acquire a handle to the destination bar in the UI. If we fail to upgrade, it's because
+            // the UI has shut down: we fail the callback to have the logging module directly log to
+            // stderr at that point.
+            let dest_bar = {
+              let stderr_dest_bar = stderr_dest_bar.lock();
+              // We can safely unwrap here because the Mutex is held until the bar is initialized.
+              stderr_dest_bar.as_ref().unwrap().upgrade().ok_or(())?
+            };
+            dest_bar.println(msg);
+            Ok(())
+          }))?
+        };
+
+        let stderr_use_color = term_stderr_write.use_color;
+        let term = console::Term::read_write_pair_with_style(
+          term_read,
+          term_stderr_write,
+          console::Style::new().force_styling(stderr_use_color),
+        );
+        // NB: We render more frequently than we receive new data in order to minimize aliasing where a
+        // render might barely miss a data refresh.
+        let draw_target = ProgressDrawTarget::term(term, ConsoleUI::render_rate_hz() * 2);
+        let multi_progress = MultiProgress::with_draw_target(draw_target);
+
+        let bars = (0..local_parallelism)
+          .map(|_n| {
+            let style = ProgressStyle::default_bar().template("{spinner} {wide_msg}");
+            multi_progress.add(ProgressBar::new(50).with_style(style))
```

Comment:
Actually... this might just "fix" #13367 if we also use the height. (Although I still think prodash is sleeker). I might try this out in a separate PR.

Reply:
Yuuuuup :joy: https://github.com/pantsbuild/pants/pull/14228

Reply:
Thanks for that!

When we first started using `indicatif`, it did not support removing bars, and so I didn't figure that it would be worth it to initialize the bar count to the terminal height if it would still break when you resized. But that's the peril of making perfect the enemy of the good.

Reply:
I've tried briefly to also do the add/remove `indicatif` lines and couldn't get it right within a reasonable amount of time.
Probably worth picking up again, but `prodash` might just be a better solution all around :man_shrugging: